src/main/java/com/android/volley/AsyncCache.java

```diff
@@ -0,0 +1,96 @@
+package com.android.volley;
+
+import androidx.annotation.Nullable;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * AsyncCache is an abstract class which implements the Cache interface for backwards compatibility,
+ * although it is not meant to be used as a Cache directly. For each method in the Cache interface
+ * (get, put, initialize, invalidate, remove, clear), it defines a corresponding abstract version
+ * which takes a Callback to be invoked upon completion instead of returning the result directly.
+ * The methods in the cache interface are implemented by making calls to the asynchronous method and
+ * blocking until they complete.
+ */
+public abstract class AsyncCache implements Cache {
+
+    public interface OnGetCompleteCallback {
+        void onGetComplete(@Nullable Entry entry);
+    }
+
+    /**
+     * Retrieves an entry from the cache and send it back through the callback function
+     *
+     * @param key Cache key
+     * @param callback Callback function that sets the retrieved entry and lets the program know it
+     *     is finished retrieving information.
+     */
+    public abstract void get(String key, OnGetCompleteCallback callback);
+
+    @Nullable
+    @Override
+    public Entry get(String key) {
+        final CountDownLatch latch = new CountDownLatch(1);
+        final AtomicReference<Entry> entryRef = new AtomicReference<>();
+        get(
+                key,
+                new OnGetCompleteCallback() {
+                    @Override
+                    public void onGetComplete(Entry entry) {
+                        entryRef.set(entry);
+                        latch.countDown();
+                    }
+                });
+        try {
+            latch.await();
+            return entryRef.get();
+        } catch (InterruptedException e) {
```

Comment:
We should probably log the exception here as well (with VolleyLog). Otherwise, the operation might not have gone through, but apart from the thread's interrupt status (which is often unchecked by callers), there'd be no visible indication that something had been canceled.

Reply:
Consider awaitUninteruptibly. Is this thread exposed enough to be interrupted? If not, this handling is just another section of untested code.

Reply:
awaitInterruptibly isn't a method on CountDownLatch.

While I agree interrupt handling is a mess in Java, I think we can at least ensure we handle them correctly to the extent possible within Volley (we do have RequestQueue#stop which would stop the executors if we control them).

Reply:
At least a Future (incl. CompletableFuture) can be canceled.